src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -64,21 +80,38 @@ export class CellExecutionQueue {
     public async cancel(forced?: boolean): Promise<void> {
         this.cancelledOrCompletedWithErrors = true;
         traceInfo('Cancel pending cells');
-        await Promise.all(this.queueOfCellsToExecute.map((item) => item.cancel(forced)));
+        await Promise.all(
+            this.queueToExecute.map((item) => {
+                if (item.cellExecution) {
+                    void item.cellExecution.cancel(forced);
+                } else {
+                    void item.codeExecution?.cancel(forced);
+                }
+            })
+        );
     }
     /**
      * Wait for cells to complete (for for the queue of cells to be processed)
      * If cells are cancelled, they are not processed, & that too counts as completion.
      * If no cells are provided, then wait on all cells in the current queue.
      */
     public async waitForCompletion(cells?: NotebookCell[]): Promise<NotebookCellRunState[]> {
+        const queue: CellExecution[] = this.getCellExecutions(this.queueToExecute);
         const cellsToCheck =
-            Array.isArray(cells) && cells.length > 0
-                ? this.queueOfCellsToExecute.filter((item) => cells.includes(item.cell))
-                : this.queueOfCellsToExecute;
+            Array.isArray(cells) && cells.length > 0 ? queue.filter((item) => cells.includes(item.cell)) : queue;
 
         return Promise.all(cellsToCheck.map((cell) => cell.result));
     }
+    public async waitForHiddenOutput(code: string): Promise<nbformat.IOutput[]> {
```

Comment:
This won't work if we have multiple expositions with the same code, e.g. `pass` or the like..
It will return the result from the first matching Execution with the same code